In [19]:
# importamos las librerias necesarias como nltk para los stopwords
import numpy as np
import pandas as pd
import re
import math
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from collections import Counter
from numpy.linalg import norm

# Carga y procesamiento de datos

In [20]:
#cargamos el dataset
df_reviews = pd.read_csv('/kaggle/input/rotten-tomatoes-movies-and-critic-reviews-dataset/rotten_tomatoes_critic_reviews.csv')
df_movies = pd.read_csv('/kaggle/input/rotten-tomatoes-movies-and-critic-reviews-dataset/rotten_tomatoes_movies.csv')

In [21]:
df_reviews

rotten_tomatoes_link        critic_name  top_critic  \
0                  m/0814255    Andrew L. Urban       False   
1                  m/0814255      Louise Keller       False   
2                  m/0814255                NaN       False   
3                  m/0814255       Ben McEachen       False   
4                  m/0814255        Ethan Alter        True   
...                      ...                ...         ...   
1130012          m/zulu_dawn      Chuck O'Leary       False   
1130013          m/zulu_dawn          Ken Hanke       False   
1130014          m/zulu_dawn    Dennis Schwartz       False   
1130015          m/zulu_dawn  Christopher Lloyd       False   
1130016          m/zulu_dawn     Brent McKnight       False   

                          publisher_name review_type review_score review_date  \
0                         Urban Cinefile       Fresh          NaN  2010-02-06   
1                         Urban Cinefile       Fresh          NaN  2010-02-06   
2                    FILMINK (Australia)       Fresh          NaN  2010-02-09   
3                Sunday Mail (Australia)       Fresh        3.5/5  2010-02-09   
4                     Hollywood Reporter      Rotten          NaN  2010-02-10   
...                                  ...         ...          ...         ...   
1130012                 Fantastica Daily      Rotten          2/5  2005-11-02   
1130013  Mountain Xpress (Asheville, NC)       Fresh        3.5/5  2007-03-07   
1130014    Dennis Schwartz Movie Reviews       Fresh           B+  2010-09-16   
1130015          Sarasota Herald-Tribune      Rotten        3.5/5  2011-02-28   
1130016             The Last Thing I See      Rotten            C  2020-07-09   

                                            review_content  
0        A fantasy adventure that fuses Greek mythology...  
1        Uma Thurman as Medusa, the gorgon with a coiff...  
2        With a top-notch cast and dazzling special eff...  
3        Whether audiences will get behind The Lightnin...  
4        What's really lacking in The Lightning Thief i...  
...                                                    ...  
1130012                                                NaN  
1130013  Seen today, it's not only a startling indictme...  
1130014  A rousing visual spectacle that's a prequel of...  
1130015  A simple two-act story: Prelude to war, and th...  
1130016  Rides the line between being a pure artifact o...  

[1130017 rows x 8 columns]

In [22]:
df_movies

rotten_tomatoes_link  \
0                                  m/0814255   
1                                  m/0878835   
2                                       m/10   
3                     m/1000013-12_angry_men   
4      m/1000079-20000_leagues_under_the_sea   
...                                      ...   
17707                            m/zoot_suit   
17708                             m/zootopia   
17709                      m/zorba_the_greek   
17710                                 m/zulu   
17711                            m/zulu_dawn   

                                             movie_title  \
0      Percy Jackson & the Olympians: The Lightning T...   
1                                            Please Give   
2                                                     10   
3                        12 Angry Men (Twelve Angry Men)   
4                           20,000 Leagues Under The Sea   
...                                                  ...   
17707                                          Zoot Suit   
17708                                           Zootopia   
17709                                    Zorba the Greek   
17710                                               Zulu   
17711                                          Zulu Dawn   

                                              movie_info  \
0      Always trouble-prone, the life of teenager Per...   
1      Kate (Catherine Keener) and her husband Alex (...   
2      A successful, middle-aged Hollywood songwriter...   
3      Following the closing arguments in a murder tr...   
4      In 1866, Professor Pierre M. Aronnax (Paul Luk...   
...                                                  ...   
17707  Mexican-American gangster Henry Reyna (Daniel ...   
17708  From the largest elephant to the smallest shre...   
17709  Traveling to inspect an abandoned mine his fat...   
17710  In 1879, the Zulu nation hands colonial Britis...   
17711  Sir Henry Bartle Frere's (John Mills) vastly o...   

                                       critics_consensus content_rating  \
0      Though it may seem like just another Harry Pot...             PG   
1      Nicole Holofcener's newest might seem slight i...              R   
2      Blake Edwards' bawdy comedy may not score a pe...              R   
3      Sidney Lumet's feature debut is a superbly wri...             NR   
4      One of Disney's finest live-action adventures,...              G   
...                                                  ...            ...   
17707                                                NaN              R   
17708  The brilliantly well-rounded Zootopia offers a...             PG   
17709                                                NaN             NR   
17710  Zulu patiently establishes a cast of colorful ...             PG   
17711                                                NaN             PG   

                                                  genres  \
0      Action & Adventure, Comedy, Drama, Science Fic...   
1                                                 Comedy   
2                                        Comedy, Romance   
3                                        Classics, Drama   
4               Action & Adventure, Drama, Kids & Family   
...                                                  ...   
17707                   Drama, Musical & Performing Arts   
17708              Action & Adventure, Animation, Comedy   
17709  Action & Adventure, Art House & International,...   
17710                                    Classics, Drama   
17711  Action & Adventure, Art House & International,...   

                                  directors  \
0                            Chris Columbus   
1                         Nicole Holofcener   
2                             Blake Edwards   
3                              Sidney Lumet   
4                         Richard Fleischer   
...                                     ...   
17707                           Luis Valdez   
17708  

In [49]:
# Usamos 'rotten_tomatoes_link' como la llave común entre ambas tablas.
df_reviews['review_content'] = df_reviews['review_content'].fillna('')

reviews_agrupadas = df_reviews.groupby('rotten_tomatoes_link')['review_content'].apply(lambda x: ' '.join(x))

# Unimos la info de películas con sus críticas
df_full = pd.merge(df_movies, reviews_agrupadas, on='rotten_tomatoes_link', how='left')

# Creamos el documento completo (Título + Género + Sinopsis + Críticas)
# Llenamos nulos con vacío "" para evitar errores
df_full['movie_content'] = (
    df_full['movie_title'].fillna('') + " " + 
    df_full['genres'].fillna('') + " " + 
    df_full['movie_info'].fillna('') + " " +
    df_full['review_content'].fillna('') 
)

In [56]:
#Funcion para procesar el texto y dejarlo "limpio"
stop_words = set(stopwords.words('english'))
def procesar_texto(texto):
    # transforma a minusculas y elimina signos especiales
    text = texto.lower()
    text = re.sub(pattern=r"<.*?>", repl='', string=text)
    text = re.sub(r"[^a-z0-9\s]", '', text) 
    text = re.sub(r'\s+', ' ', text).strip()

    #dividimos el texto en tokens
    tokens = text.split()
    #eliminamos las stopwords
    tokens_filtrados = [t for t in tokens if t not in stop_words and len(t) > 2]
    
    return tokens_filtrados


In [57]:
# aplicamos la función de procesar datos al corpus original y le transformamos a una lista ordenada 
corpus = df_full['movie_content'].apply(procesar_texto).tolist()

In [51]:
corpus[1]

['please',
 'give',
 'comedy',
 'kate',
 'catherine',
 'keener',
 'husband',
 'alex',
 'oliver',
 'platt',
 'wealthy',
 'new',
 'yorkers',
 'prowl',
 'estate',
 'sales',
 'make',
 'tidy',
 'profit',
 'reselling',
 'items',
 'bought',
 'cheaply',
 'buy',
 'apartment',
 'next',
 'door',
 'plan',
 'remodel',
 'soon',
 'current',
 'occupant',
 'cranky',
 'old',
 'woman',
 'dies',
 'kate',
 'troubled',
 'way',
 'husband',
 'earn',
 'living',
 'tries',
 'assuage',
 'guilt',
 'befriending',
 'tenant',
 'womans',
 'granddaughters',
 'overtures',
 'lead',
 'unexpected',
 'consequences',
 'like',
 'holofceners',
 'previous',
 'pictures',
 'please',
 'give',
 'derives',
 'narrative',
 'energy',
 'less',
 'series',
 'plotted',
 'incidents',
 'keenly',
 'observed',
 'interplay',
 'clashing',
 'personality',
 'tics',
 'worldviews',
 'muted',
 'almost',
 'rohmerlike',
 'moral',
 'tale',
 'doesnt',
 'quite',
 'dive',
 'deeply',
 'enough',
 'several',
 'characters',
 'keener',
 'peet',
 'hall',
 'shine

# Contrucción del sistema BM25

In [52]:
# obtenemos las estadisticas que usamos en la formula de bm25 que son el tamaño de cada documento
# y el promedio del tamaño de los documentos del corpus
def obtener_estadisticas_bm25(corpus):
    doc_lengths = [len(doc) for doc in corpus]
    avgdl = sum(doc_lengths) / len(corpus)
    return doc_lengths, avgdl

In [53]:
# calculamos idf para bm25
def calcular_idf_bm25(corpus):
    N = len(corpus)
    idf_dict = {}
    df_counts = Counter()
    for doc in corpus:
        df_counts.update(set(doc))
    
    for palabra, df in df_counts.items():
        idf_dict[palabra] = math.log((N - df + 0.5) / (df + 0.5) + 1)
    return idf_dict

In [58]:
# funcion que calcula el score para cada documento 
# ademas asignamos un valor de k1=1.5 para que la repetición de palabras no inflara el score
# b = 0.75 para que los documentos extensos no dominen frente a otros documentos con menos palabras
def score_bm25_documento(query_tokens, doc_tokens, doc_len, avgdl, idf_dict, k1=1.5, b=0.75):
    score = 0
    doc_counts = Counter(doc_tokens)
    
    for token in query_tokens:
        if token in doc_counts:
            tf = doc_counts[token]
            idf = idf_dict.get(token, 0)
            numerador = tf * (k1 + 1)
            denominador = tf + k1 * (1 - b + b * (doc_len / avgdl))
            score += idf * (numerador / denominador)
    return score

In [48]:
def buscar_bm25(consulta, corpus, top_n=5, k1=1.5, b=0.75):
    # Procesamos consulta y la transformamos a tokens
    query_tokens = procesar_texto(consulta)
    
    # obtenemos las estadisticas para bm25
    doc_lengths, avgdl = obtener_estadisticas_bm25(corpus)
    idf_bm25 = calcular_idf_bm25(corpus)
    
    # Calculo de los scores para todos los documentos
    scores = []
    for i, doc_tokens in enumerate(corpus):
        puntos = score_bm25_documento(
            query_tokens, doc_tokens, doc_lengths[i], avgdl, idf_bm25, k1, b
        )
        if puntos > 0:
            # Aquí guardamos id, puntos
            scores.append((i, puntos))

    ranking = sorted(scores, key=lambda x: x[1], reverse=True)[:top_n]
    
    return ranking

# funcon que permite realizar la busqueda pasandole la consulta e imprime los resultados con los valores de bm25
# ademas muestra información como el titulo, genero y la sipnosis para visualizar si los resultados son correctos
def motor_de_busqueda(consulta):
    print(f"Buscando: '{consulta}'")

    ranking = buscar_bm25(consulta, corpus)
    
    resultados = []

    for doc_id, score in ranking:
        
        # Recuperamos los datos originales usando el ID
        pelicula = df_full.iloc[doc_id]
        
        resultados.append({
            'Título': pelicula['movie_title'],
            'Score BM25': round(score, 4),
            'Género': pelicula['genres'],
            # Convertimos a string por seguridad en caso de nulos
            'Sinopsis': str(pelicula['movie_info'])[:150] + "..." 
        })
        
    if not resultados:
        return "No se encontraron coincidencias."
        
    return pd.DataFrame(resultados)

# Pruebas y resultados

In [54]:
print("\n--- RESULTADOS PARA 'VIAJES ESPACIALES' ---")
display(motor_de_busqueda("Space travel movies"))


--- RESULTADOS PARA 'VIAJES ESPACIALES' ---
Buscando: 'Space travel movies'


Título  Score BM25  \
0                         Hubble 3D     12.3587   
1      Futurama: Bender's Big Score     12.0163   
2  It! The Terror From Beyond Space     11.5168   
3                Fly Me to the Moon     11.3651   
4                 Somewhere in Time     11.3624   

                                              Género  \
0                                 Documentary, Drama   
1       Animation, Comedy, Science Fiction & Fantasy   
2        Classics, Horror, Science Fiction & Fantasy   
3  Animation, Kids & Family, Science Fiction & Fa...   
4          Drama, Science Fiction & Fantasy, Romance   

                                            Sinopsis  
0  In May 2009, NASA astronauts embark on a missi...  
1  Alien invaders force Bender (John DiMaggio) th...  
2  It's the future -- 1973 -- and a rescue team i...  
3  Nat the fly loves listening to his grandfather...  
4  In 1972, playwright Richard Collier (Christoph...

In [55]:
print("\n--- RESULTADOS PARA 'PELICULAS FAMILIARES' ---")
display(motor_de_busqueda("Family movies for kids"))


--- RESULTADOS PARA 'PELICULAS FAMILIARES' ---
Buscando: 'Family movies for kids'


Título  Score BM25  \
0                               Spy Kids      8.7466   
1                                Snowmen      8.5936   
2                 The Kids Are All Right      8.4686   
3     Oddball and the Penguins (Oddball)      8.4059   
4  Spy Kids 2: The Island of Lost Dreams      8.3152   

                                              Género  \
0  Action & Adventure, Comedy, Kids & Family, Mys...   
1                               Drama, Kids & Family   
2                                      Comedy, Drama   
3  Action & Adventure, Art House & International,...   
4  Action & Adventure, Kids & Family, Mystery & S...   

                                            Sinopsis  
0  Two young kids become spies in attempt to save...  
1  A 10-year-old boy and his friends perform peri...  
2  Lesbian couple Jules (Julianne Moore) and Nic ...  
3  A chicken farmer and his granddaughter train a...  
4  Now full fledged Spy Kids, Carmen (Alexa Vega)...

In [59]:
print("\n--- RESULTADOS PARA BUSQUEDA EXTRA ---")
display(motor_de_busqueda("Action movies in space"))


--- RESULTADOS PARA BUSQUEDA EXTRA' ' ---
Buscando: 'Action movies in space'


Título  Score BM25  \
0                                      Lost in Space      8.9624   
1                                      Space Cowboys      8.9566   
2  Ôritsu uchûgun Oneamisu no tsubasa (Royal Spac...      8.9413   
3                                       Space Chimps      8.8707   
4                         Zathura: A Space Adventure      8.8207   

                                              Género  \
0      Action & Adventure, Science Fiction & Fantasy   
1  Action & Adventure, Comedy, Drama, Science Fic...   
2  Animation, Anime & Manga, Drama, Science Ficti...   
3  Action & Adventure, Animation, Comedy, Kids & ...   
4  Action & Adventure, Kids & Family, Science Fic...   

                                            Sinopsis  
0  The space family Robinson is on a journey to A...  
1  Clint Eastwood, Tommy Lee Jones, Donald Suther...  
2  In this animated film, Shiro (Leo Morimoto) fa...  
3  Ham III (Andy Samberg) is the grandson of the ...  
4  After their father (Tim Robbins) is called int...

# Análisis de resultados

## Evaluación de la Consulta 1: "Space travel movies"
Para esta consulta, el sistema BM25 mostró una alta precisión, devolviendo documentos con scores de relevancia entre 11.36 y 12.35.

El sistema recuperó desde animación (Futurama) hasta terror clásico (It! The Terror From Beyond Space). Esto indica que el buscador no está sesgado hacia un solo género, sino que rastrea el concepto "viaje espacial" a través de todo el corpus.

La película Somewhere in Time (Puesto 5) trata sobre viajes en el tiempo, no espaciales. Este es un caso interesante de coincidencia parcial, el algoritmo encontró las palabras "travel" y "movies" con alta frecuencia, pero carece de la comprensión semántica para distinguir "tiempo" de "espacio".

## Evaluación de la Consulta 2: "Family movies for kids"
Los scores para esta consulta fueron ligeramente más bajos (rango 8.31 - 8.74), lo que indica que los términos son más comunes en la colección y por tanto tienen un IDF menor.

El Top 1 y Top 5 son películas de la saga "Spy Kids". Aquí el sistema funcionó a la perfección: coincidió la palabra "Kids" en el título, el género "Kids & Family" y seguramente múltiples menciones en las críticas.

El resultado #3 es "The Kids Are All Right". Aunque el algoritmo la trajo por contener las palabras "Kids" y probablemente "Family", esta es una película de Drama/Comedia orientada a adultos, no una película infantil.

## Validación del Modelo BM25 y la Fusión de Datos

La implementación de BM25 con parámetros $k1=1.5$ y $b=0.75$ fue crucial para el éxito del sistema frente a un modelo TF-IDF tradicional, fusionar todas las críticas (critic_reviews) con la sinopsis, creamos documentos de longitudes extremas. Sin la normalización de BM25, las películas populares con miles de palabras hubieran dominado el ranking simplemente por probabilidad. El parámetro $b$ corrigió este sesgo. En las críticas fusionadas, palabras como "movie" o "story" se repiten cientos de veces. La saturación de BM25 evitó que estos términos comunes inflaran artificialmente los puntajes.
## Propuesta de Mejoras Futuras 

Para solucionar el error de "The Kids Are All Right", se debería incorporar un filtro para que coincida con la busqueda.
Para la búsqueda de "Space travel", el sistema podría no encontrar películas que solo usen términos como "Interstellar" o "Orbit" sin decir explícitamente "travel".
Dado que la base de datos está en inglés, el sistema actual requiere consultas en inglés. Se podria implementar una herramienta que permita escribir busquedas en español e ingles.